python/taulabs/uavo.py

```diff
@@ -183,18 +212,64 @@ def make_class(xml_file):
                     info['elements'] += 1
 
             if info['type'] == 'enum':
-                info['options'] = []
+                info['options'] = {}
                 if field.get('options'):
                     # we've got an inline "options" attribute
-                    for option_text in field.get('options').split(','):
-                        info['options'].append(option_text.strip(' '))
+                    for ii, option_text in enumerate(field.get('options').split(',')):
+                        option_text = RE_SPECIAL_CHARS.sub('', option_text)
+                        info['options'][option_text] = ii
                 else:
                     # we must have some 'option' elements in this sub-tree
-                    for option_text in [option.text for option in field.findall('options/option')]:
-                        info['options'].append(option_text)
+                    for ii, option_text in enumerate([option.text
+                            for option in field.findall('options/option')]):
+                        option_text = RE_SPECIAL_CHARS.sub('', option_text)
+                        info['options'][option_text] = ii
 
             # convert type string to an int
             info['type_val'] = type_enum_map[info['type']]
+
+            # Get parent
+            if info['parent'] is not None:
+                parent_name = info['parent'].split('.')
+                field_name = RE_SPECIAL_CHARS.sub('', parent_name[1])
+                try:
+                    parent_class = globals()['UAVO_' + parent_name[0]]
+                    parent_options = getattr(parent_class, 'ENUM_' + field_name)
+                    if len(info['options']) == 0:
```

Comment:
The hash for whenever there's a sparse enum includes the next index before there's a gap.  So it only affects the hw* objects.

```
        if(info->fields[n]->type == FIELDTYPE_ENUM) {
            QStringList options = info->fields[n]->options;
            int nextIdx = 0;

            for (int m = 0; m < options.length(); m++) {
                int idx = findOptionIndex(info->fields[n], m);

                if (idx < 0) abort();

                // Not contiguous options.  Update with next value.
                if (idx != nextIdx) {
                    hash = updateHash((quint32) idx, hash);
                }

                nextIdx = idx+1;

                hash = updateHash(options[m], hash);
            }
```

is how it works on the C side.  now that you understand enums it's possible to fix.   It's issue #1804